packages/next-swc/crates/core/src/emotion/mod.rs

```diff
@@ -0,0 +1,842 @@
+    Css,
+    Styled,
+    GlobalJSX,
+}
+
+impl Default for ExprKind {
+    fn default() -> Self {
+        ExprKind::Css
+    }
+}
+
+#[derive(Debug)]
+enum PackageMeta {
+    Named(ExprKind),
+    Namespace(EmotionModuleConfig),
+}
+
+pub fn emotion<'a, C: Comments>(
+    emotion_options: EmotionOptions,
+    path: &Path,
+    cm: Arc<SourceMap>,
+    comments: &'a C,
+) -> impl Fold + 'a {
+    EmotionTransformer::new(emotion_options, path, cm, comments)
+}
+
+pub struct EmotionTransformer<'a, C: Comments> {
+    pub options: EmotionOptions,
+    filepath_hash: Option<u32>,
+    filepath: PathBuf,
+    dir: Option<String>,
+    filename: Option<String>,
+    cm: Arc<SourceMap>,
+    comments: &'a C,
+    import_packages: FxHashMap<String, PackageMeta>,
+    emotion_target_class_name_count: usize,
+    current_context: Option<String>,
+    // skip `css` transformation if it in JSX Element/Attribute
+    in_jsx_element: bool,
+}
+
+impl<'a, C: Comments> EmotionTransformer<'a, C> {
+    pub fn new(options: EmotionOptions, path: &Path, cm: Arc<SourceMap>, comments: &'a C) -> Self {
+        EmotionTransformer {
+            options,
+            filepath_hash: None,
+            filepath: path.to_owned(),
+            dir: path.parent().and_then(|p| p.to_str()).map(|s| s.to_owned()),
+            filename: path
+                .file_name()
+                .and_then(|filename| filename.to_str())
+                .map(|s| s.to_owned()),
+            cm,
+            comments,
+            import_packages: FxHashMap::default(),
+            emotion_target_class_name_count: 0,
+            current_context: None,
+            in_jsx_element: false,
+        }
+    }
+
+    #[inline]
+    // Compute file hash on demand
+    // Memorize the hash of the file name
+    fn get_filename_hash(&mut self) -> u32 {
+        if self.filepath_hash.is_none() {
+            self.filepath_hash = Some(hash::murmurhash2(
+                self.filepath.to_string_lossy().as_bytes(),
+                0,
+            ));
+        }
+        self.filepath_hash.unwrap()
+    }
+
+    fn create_label(&self, with_prefix: bool) -> String {
+        let prefix = if with_prefix { "label:" } else { "" };
+        let mut label = format!(
+            "{}{}",
+            prefix,
+            self.options
+                .label_format
+                .clone()
+                .unwrap_or_else(|| "[local]".to_owned())
+        );
+        if let Some(current_context) = &self.current_context {
+            label = label.replace("[local]", current_context);
+            if let Some(filename) = self.filename.as_ref() {
+                label = label.replace("[filename]", filename);
+            }
+            if let Some(dir) = self.dir.as_ref() {
+                label = label.replace("[dir]", dir);
+            };
+        }
+        label
+    }
+
+    fn create_sourcemap(&mut self, pos: BytePos) -> Option<String> {
+        if self.options.sourcemap.unwrap_or(false) {
```

Comment:
It's for development purposes, and it's one of the major features in `@emotion/babel-plugin`. 
The sourcemap generated here will be inserted into `<style></style>` tag in HTML, so that you can jump to the source from computed styles:

https://user-images.githubusercontent.com/3468483/157389915-6f08dc9b-f3b2-47b3-b9cf-6768681f81fe.mp4

Reply:
Sourcemap will be disabled if the build type is `production`.

Reply:
Wow I never thought this is possible.
Thank you for explaining!